subsys/bluetooth/audio/ascs.c

```diff
@@ -1496,41 +1507,187 @@ static bool ascs_codec_store_metadata(struct bt_data *data, void *user_data)
+	case BT_AUDIO_METADATA_TYPE_EXTENDED: /* 1 - 255 octets */
+	case BT_AUDIO_METADATA_TYPE_VENDOR: /* 1 - 255 octets */
+		if (len < 1) {
+			return false;
+		}
+
+		return true;
+	case BT_AUDIO_METADATA_TYPE_CCID_LIST: /* 2 - 254 octets */
+		if (len < 2) {
+			return false;
+		}
+
+		return true;
+	case BT_AUDIO_METADATA_TYPE_PROGRAM_INFO: /* 0 - 255 octets */
+	case BT_AUDIO_METADATA_TYPE_PROGRAM_INFO_URI: /* 0 - 255 octets */
+		return true;
+	default:
 		return false;
 	}
+}
 
-	meta = &codec->meta[codec->meta_count];
+static bool ascs_parse_metadata(struct bt_data *data, void *user_data)
+{
+	struct ascs_parse_result *result = user_data;
+	const struct bt_audio_ep *ep = result->ep;
+	const uint8_t data_len = data->data_len;
+	const uint8_t data_type = data->type;
+	const uint8_t *data_value = data->data;
+
+	result->count++;
+
+	BT_DBG("#%u type 0x%02x len %u", result->count, data_type, data_len);
+
+	if (result->count > CONFIG_BT_CODEC_MAX_METADATA_COUNT) {
+		BT_ERR("Not enough buffers for Codec Config Metadata: %zu > %zu",
+		       result->count, CONFIG_BT_CODEC_MAX_METADATA_LEN);
+		result->err = -ENOMEM;
 
-	if (data->data_len > sizeof(meta->value)) {
-		BT_ERR("Not enough space for Codec Config Metadata: %u > %zu",
-		       data->data_len, sizeof(meta->value));
 		return false;
 	}
 
-	BT_DBG("#%u type 0x%02x len %u", codec->meta_count, data->type,
-	       data->data_len);
+	if (!ascs_valid_metadata_type(data_type, data_len)) {
+		result->err = -EINVAL;
 
-	meta->data.type = data->type;
-	meta->data.data_len = data->data_len;
+		return false;
+	}
 
-	/* Deep copy data contents */
-	meta->data.data = meta->value;
-	(void)memcpy(meta->value, data->data, data->data_len);
+	if (data_len > CONFIG_BT_CODEC_MAX_METADATA_LEN) {
+		BT_ERR("Not enough space for Codec Config Metadata: %u > %zu",
+		       data->data_len, CONFIG_BT_CODEC_MAX_METADATA_LEN);
+		result->err = -ENOMEM;
+
+		return false;
+	}
+
+	/* The CAP acceptor shall not accept metadata with
+	 * unsupported stream context.
+	 */
+	if (IS_ENABLED(CONFIG_BT_CAP_ACCEPTOR)) {
+		if (data_type == BT_AUDIO_METADATA_TYPE_STREAM_CONTEXT) {
+			const uint16_t context = sys_get_le16(data_value);
 
-	BT_HEXDUMP_DBG(meta->value, data->data_len, "data");
+			if (!bt_pacs_context_available(ep->dir, context)) {
+				result->err = -EACCES;
 
-	codec->meta_count++;
+				return false;
+			}
+		} else if (data_type == BT_AUDIO_METADATA_TYPE_CCID_LIST) {
+			/* Verify that we the CCID is a known CCID on the
+			 * writing device
+			 */
+			for (uint8_t i = 0; i < data_len; i++) {
+				const uint8_t ccid = data_value[i];
+
+				if (!bt_cap_acceptor_ccid_exist(ep->stream->conn,
+								ccid)) {
+					BT_WARN("CCID %u is unknown", ccid);
+
+					/* TBD:
+					 * Should we reject the Metadata?
+					 *
+					 * Should unknown CCIDs trigger a
+					 * discovery procedure for TBS or MCS?
+					 *
+					 * Or should we just accept as is, and
+					 * then let the application decide?
+					 */
```

Comment:
Maybe nothing to-do, but a comment on this comment, I have noticed clients that does send the CCID list seemingly based only on supporting conversational context, even when TBS or MCS is not supported

Reply:
It does not make any sense to put a CCID if TBS or MCS is not supported. CCID is used to identify a TBS or MCS instance. The CCID has no context or value outside of that.

Reply:
I'm in full agreement with you :)